File: algorithms/2_medium/T/01996_the_number_of_weak_characters_in_the_game/sort.c
__attribute__((no_sanitize("all")))
static int cmp(const void *restrict a_ptr, const void *restrict b_ptr) {
    const int *const restrict a = *(const int *const restrict *const restrict)a_ptr;
    const int *const restrict b = *(const int *const restrict *const restrict)b_ptr;

    if (a[0] == b[0]) {
        return (a[1] > b[1]) - (a[1] < b[1]);
    } else {
        return (a[0] > b[0]) - (a[0] < b[0]);
    }
}

__attribute__((no_sanitize("all")))
int numberOfWeakCharacters(int *restrict *restrict properties, const int len, const int *restrict cols_data) {
    if (len < 2 || properties == NULL || cols_data == NULL || cols_data[0] != 2) { return -1; }

    qsort(properties, (size_t)len, sizeof(*properties), cmp);

    int cnt = 0;
    int curr_max = INT_MIN;
    int level = properties[len - 1][0];
    int level_max = properties[len - 1][1];

    for (int i = len - 1; i >= 0; i -= 1) {
        const int *restrict p = properties[i];
        if (p[0] < level) {
            curr_max = MAX(curr_max, level_max);
            level = p[0];
            level_max = p[1];
        }

        if (p[1] < curr_max) {
            cnt += 1;
        }
    }

    return cnt;
}

File: algorithms/2_medium/T/01996_the_number_of_weak_characters_in_the_game/sort.rs
impl Solution {
    pub fn number_of_weak_characters(mut properties: Vec<Vec<i32>>) -> i32 {
        assert!(properties.len() >= 2);

        properties.sort_unstable_by(|a, b| {
            if a[0] == b[0] {
                a[1].cmp(&b[1])
            } else {
                a[0].cmp(&b[0])
            }
        });

        let mut cnt = 0_usize;
        let mut curr_max = i32::MIN;
        let mut level = properties.last().unwrap()[0];
        let mut level_max = properties.last().unwrap()[1];

        for p in properties.into_iter().rev() {
            if p[0] < level {
                curr_max = curr_max.max(level_max);
                level = p[0];
                level_max = p[1];
            }

            if p[1] < curr_max {
                cnt += 1;
            }
        }

        cnt.try_into().unwrap()
    }
}
